src/reconstruction/community.jl

```diff
@@ -0,0 +1,409 @@
+"""
+add_objective!(community::CoreModel, objective_mets::Vector{String}; objective_weights=Float64[], objective_column_index=0)
+
+Add an objective to the `community` model. Supply the string names of the
+objective metabolites in `objective_mets`. Optionally specify the weight to
+assign each metabolite in the objective function, if unassigned then equal
+weight is assumed. Also, optionally specify whether the objective already exists
+in the model by assigning `objective_column_index`. If unassigned then an
+objective column will be added, otherwise the column at `objective_column_index`
+will be updated.
+
+Note, the weights are negated inside the function so that the objective metabolites 
+are seen as reagents/substrates, not products in the reaction equation. 
+
+# Example
+```
+add_objective!(model, ["met1", "met2"]) # adds a new column with weights = [1,1]
+add_objective!(model, ["met1", "met2"]; objective_weights=[0.1, 0.9]) # adds a new column
+add_objective!(model, ["met1", "met2"]; objective_weights=[0.1, 0.9], objective_column_index=10) # updates column 10
+```
+"""
+function add_objective!(
+    community::CoreModel,
+    objective_mets::Vector{String};
+    objective_weights = Float64[],
+    objective_column_index = 0,
+)
+    obj_inds = indexin(objective_mets, metabolites(community))
+    if isempty(objective_weights)
+        objective_weights = repeat([1.0], inner = length(objective_mets))
+    end
+
+    if objective_column_index == 0 # needs to be created
+        nr, _ = size(community.S)
+        objcol = spzeros(nr)
+        objcol[obj_inds] .= -objective_weights
+
+        # extend model by one reaction 
+        community.S = hcat(community.S, objcol)
+        community.xl = [community.xl; 0.0]
+        community.xu = [community.xu; 1000.0]
+        community.rxns = [community.rxns; "community_biomass"]
+        community.c = spzeros(size(community.S, 2))
+        community.c[end] = 1.0
+    else # only update
+        community.S[:, objective_column_index] .= 0.0 # reset
+        community.S[obj_inds, objective_column_index] .= -objective_weights
+        dropzeros!(community.S)
+        community.c = spzeros(size(community.S, 2))
+        community.c[objective_column_index] = 1.0
+        community.xl[objective_column_index] = 0.0
+        community.xu[objective_column_index] = 1000.0
+    end
+end
+
+add_objective!(model::CoreModel, objective_met::String, objective_weight::Float64) =
+    add_objective!(model, [objective_met], [objective_weight])
+
+"""
+add_model(
```

Comment:
spawning `M` into the type is not required, you don't need to unify it anywhere. Just do ::MetabolicModel.

Do not use empty strings (empty string is a perfectly valid ID) -- you want `nothing`.

Include the return type.

The thing soudns a bit like it could make a modification but the semantics (and arguments) seems more like inplace, choose either of that.

Reply:
(I'd vote for inplace here)

Reply:
Hmm this function is not really in place... It deconstructs everything and then makes a new model. I am not sure how efficient and in place version will be given that you need to extend all the matrices (you may as well just make a new one)?

Or it can be made trivially in place by just overwriting the community input model?